### PR TITLE
Issue/21 - Introduce query_var_default_value and is_query_var_default().

### PR DESCRIPTION
**Pull request by JJJ**
First pass at allowing querying explicitly for falsy values `(0, '0', false, null)`

This work-in-progress is a proof-of-concept at how we could potentially avoid bugs with `empty()` checks against `query_vars`.

See #21.

### CODE REVIEW

**Comment by JJJ** [comment]
I think this is worth checking out at this point.

The random default query_var value could (and probably should) be used in a few other places where `empty()` checks are done against `''` values. Before doing that, I figured I'd wait for feedback on what's here.

The reason I think a random value is necessary, is to avoid having a _default_ value that might unintentionally be an _actual_ value someone would want to query the database for.

By using a random value and comparing against it (instead of an empty default string) row values such as `''`, `0`, `"0"`, `false` and `null` should now be able to be queried for literally.

**Comment by JJJ** [comment]
Hey reviewers! 👋

Would love some eyes and feedback on this one whenever you have an opportunity. ❤️

- [ ] Does it actually address the issue?
- [ ] Is this approach easy enough to understand?
- [ ] Is this the best way to solve this problem?

**Comment by alexstandiford** [comment]
Hey @JJJ I went ahead and implemented this since I ran into a good test case to run this against. I ran about 200 unit tests on a personal project of mine. Of those 200 tests, a small handful were failing because of this exact issue.

__Does it actually resolve the problem__
The system resolved my problem beautifully. All of the issues that previously failed passed, and new tests did not fail because of these changes.

__Is this approach easy enough to understand__
Yep, I didn't have trouble following the intent.

__Is this the best way to solve the problem?__
Honestly, I'm not sure. In my testing, I also tried out the methodology mentioned in your comment here https://github.com/berlindb/core/issues/21#issuecomment-602908002

In this particular circumstance, it worked just as well, and I found it to be a bit easier to read. Are there situations where an `isset` check could actually fail that I'm missing?

The only (minor) concern I have is _what happens if the random value I want to save happens to be the same value as the random default?_ I know this is unlikely, but it _could_ happen. Are the odds so astronomically low that this is not a concern?

File: query.php [hunk 1]
```diff
@@ -640,6 +656,18 @@ public function set_query_var( $key = '', $value = '' ) {
...
 
+	/**
+	 * Check whether a query variable strictly equals the unique default
+	 * starting value.
+	 *
+	 * @since 1.1.0
+	 * @param string $key
+	 * @return bool
+	 */
+	public function is_query_var_default( $key = '' ) {
```
Comment (alexstandiford) [inline query.php:667]: Was this intentionally left public? I'm not opposed to that, but most other helper methods in this class are private, so I'm a little surprised this one is not.

**Comment by JJJ** [comment]
> Honestly, I'm not sure. In my testing, I also tried out the methodology mentioned in your comment here #21 (comment)

Your approach there is logical.

> In this particular circumstance, it worked just as well, and I found it to be a bit easier to read. Are there situations where an isset check could actually fail that I'm missing?

That's a great question. Technically, yes! This PR takes your suggestion and applies it to all default values, not just falsy ones, so the same collision avoidance logic is now universal.


> The only (minor) concern I have is what happens if the random value I want to save happens to be the same value as the random default? I know this is unlikely, but it could happen. Are the odds so astronomically low that this is not a concern?

This is always possible, but yes... astronomically low. We're talking Rick and Morty multiple identical dimensions running the exact same code and still getting different results kinda low. Enough monkeys could potentially type out the same random value after like a billion years. 🐵

File: query.php (hunk 1 above)
Comment (alexstandiford) [reply query.php:667]: I slept on this - and I think it makes _a lot_ of sense for this to _be_ public. It empowers developers to use the same methodology used in the class to make these false-y checks.

**Comment by alexstandiford** [comment]
> That's a great question. Technically, no. But also kinda! This PR takes your suggestion and applies it to all default values, not just falsy ones, so the same collision avoidance logic is now universal.

Oh, that makes a lot of sense!

**Comment by JJJ** [comment]
Removing reviews from Drew and Ashley to lighten the load here.

Thanks @alexstandiford for the eyes.

Merge imminent.

**Pull request merged**